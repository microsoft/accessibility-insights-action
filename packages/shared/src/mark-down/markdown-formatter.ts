// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { brand, brandLogoImg, toolName } from '../content/strings';

export const escaped = (markdown: string): string => {
    return markdown.replace(/</g, '\\<'); // CodeQL [SM02383] False Positive: The expected input does not include a backslash, so the replacement is not needed.
};

export const snippet = (text: string): string => {
    return `\`${text}\``;
};

export const link = (href: string, text: string): string => {
    return `[${text}](${href})`;
};

export const image = (altText: string, src: string): string => {
    return `![${altText}](${src})`;
};

export const listItem = (text: string): string => {
    return `* ${text}`;
};

export const heading = (text: string, headingLevel: number): string => {
    return `${'#'.repeat(headingLevel)} ${text}`;
};

export const bold = (text: string): string => {
    return `**${text}**`;
};

export const productTitle = (): string => {
    return `${image(`${brand}`, brandLogoImg)} ${toolName}`;
};

export const footerSeparator = (): string => `---`;

export const sectionSeparator = (): string => '\n';
