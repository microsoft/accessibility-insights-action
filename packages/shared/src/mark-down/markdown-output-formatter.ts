// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { injectable } from 'inversify';
import { brand, brandLogoImg, toolName } from '../content/strings';
import { OutputFormatter } from '../output/output-formatter';
@injectable()
export class MarkdownOutputFormatter implements OutputFormatter {
    public escaped(text: string): string {
        return text.replace(/</g, '\\<');
    }
    public snippet(text: string): string {
        return `\`${text}\``;
    }
    public link(href: string, text: string): string {
        return `[${text}](${href})`;
    }
    public image(altText: string, src: string): string {
        return `![${altText}](${src})`;
    }
    public listItem(text: string): string {
        return `* ${text}`;
    }
    public heading(text: string, headingLevel: number): string {
        return `${'#'.repeat(headingLevel)} ${text}`;
    }
    public bold(text: string): string {
        return `**${text}**`;
    }
    public productTitle(): string {
        return `${this.image(`${brand}`, brandLogoImg)} ${toolName}`;
    }
    public footerSeparator(): string {
        return `---`;
    }
    public sectionSeparator(): string {
        return '\n';
    }
}
