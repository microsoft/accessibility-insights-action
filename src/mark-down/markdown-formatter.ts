// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { brand, brandLogoImg, toolName } from '../content/strings';

export const snippet = (text: string): string => {
    return `\`${text}\``;
};

export const link = (href: string, text: string): string => {
    return `[${text}](${href})`;
};

export const image = (altText: string, src: string): string => {
    return `![${altText}](${src})`;
};

export const listItem = (text: string): string => {
    return `* ${text}`;
};

export const heading = (text: string, headingLevel: number): string => {
    return `${'#'.repeat(headingLevel)} ${text}`;
};

export const bold = (text: string): string => {
    return `**${text}**`;
};

export const productTitle = (): string => {
    return `${image(`${brand}`, brandLogoImg)} ${toolName}`;
};

export const footerSeparator = (): string => `---`;

export const sectionSeparator = (): string => '\n';
