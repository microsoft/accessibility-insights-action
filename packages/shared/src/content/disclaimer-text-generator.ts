// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { inject, injectable } from 'inversify';
import { MarkdownOutputFormatter } from '../mark-down/markdown-formatter';
import { toolName, webToolName } from './strings';

@injectable()
export class DisclaimerTextGenerator {
    constructor(@inject(MarkdownOutputFormatter) private readonly markdownOutputFormatter: MarkdownOutputFormatter) {}

    private assessmentLink = this.markdownOutputFormatter.link(
        'https://accessibilityinsights.io/docs/en/web/getstarted/assessment',
        'Assessments',
    );
    private webToolLink = this.markdownOutputFormatter.link('https://accessibilityinsights.io/docs/en/web/overview', webToolName);
    private wcag21AALink = this.markdownOutputFormatter.link(
        'https://www.w3.org/WAI/WCAG21/quickref/?currentsidebar=%23col_customize&levels=aaa',
        'WCAG 2.1 AA',
    );

    public generateDisclaimerText(): string {
        return `The ${toolName} runs a set of automated checks to help find some of the most common accessibility issues. The automated checks can detect accessibility problems such as missing or invalid properties, but most accessibility problems can only be discovered through manual testing.\n\nWe recommend automated testing, to continuously protect against some common issues, and regular ${this.assessmentLink} using ${this.webToolLink}, a free and open source tool that helps you assess your website or web app for ${this.wcag21AALink} coverage.`;
    }
}
