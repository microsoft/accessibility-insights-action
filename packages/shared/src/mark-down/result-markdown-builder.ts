// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { CombinedReportParameters } from 'accessibility-insights-report';
import { inject, injectable } from 'inversify';
import { ArtifactsInfoProvider } from '../artifacts-info-provider';
import { BaselineInfo } from '../baseline-info';
import { BaselineEvaluation } from 'accessibility-insights-scan';
import { brand } from '../content/strings';
import { bold, escaped, footerSeparator, heading, link, listItem, productTitle, sectionSeparator } from './markdown-formatter';
import { iocTypes } from '../ioc/ioc-types';

@injectable()
export class ResultMarkdownBuilder {
    constructor(@inject(iocTypes.ArtifactsInfoProvider) private readonly artifactsInfoProvider: ArtifactsInfoProvider) {}

    public buildErrorContent(): string {
        const lines = [
            this.headingWithMessage('Something went wrong'),
            sectionSeparator(),
            `You can review the log to troubleshoot the issue. Fix it and re-run the workflow to run the automated accessibility checks again.`,
            sectionSeparator(),
        ];

        return this.scanResultDetails(lines.join(''));
    }

    public buildContent(combinedReportResult: CombinedReportParameters, title?: string, baselineInfo?: BaselineInfo): string {
        const passedChecks = combinedReportResult.results.resultsByRule.passed.length;
        const inapplicableChecks = combinedReportResult.results.resultsByRule.notApplicable.length;
        const failedChecks = combinedReportResult.results.resultsByRule.failed.reduce((a, b) => a + b.failed.length, 0);

        let lines = [
            failedChecks === 0 ? this.headingWithMessage('All applicable checks passed') : this.headingWithMessage(),
            sectionSeparator(),
            this.urlsListItem(
                combinedReportResult.results.urlResults.passedUrls,
                combinedReportResult.results.urlResults.unscannableUrls,
                combinedReportResult.results.urlResults.failedUrls,
            ),
            sectionSeparator(),
            this.rulesListItem(passedChecks, inapplicableChecks, failedChecks),
            sectionSeparator(),

            this.downloadArtifacts(),
            this.failureDetails(combinedReportResult),
        ];

        // baselining is available
        if (baselineInfo !== undefined) {
            lines = [
                this.headingWithMessage(),
                sectionSeparator(),
                this.failureDetailsBaseline(combinedReportResult, baselineInfo),
                sectionSeparator(),
                this.baselineDetails(baselineInfo),
                sectionSeparator(),
                sectionSeparator(),
                this.downloadArtifactsWithLink(combinedReportResult, baselineInfo.baselineEvaluation),
                sectionSeparator(),
                sectionSeparator(),
                footerSeparator(),
                sectionSeparator(),
                heading('Scan summary', 4),
                sectionSeparator(),
                this.urlsListItemBaseline(
                    combinedReportResult.results.urlResults.passedUrls,
                    combinedReportResult.results.urlResults.unscannableUrls,
                    combinedReportResult.results.urlResults.failedUrls,
                ),
                sectionSeparator(),
                this.rulesListItemBaseline(passedChecks, inapplicableChecks, failedChecks),
                sectionSeparator(),
            ];
        }

        if (title !== undefined) {
            lines = [heading(title, 3), sectionSeparator()].concat(lines);
        }

        return this.scanResultDetails(lines.join(''), this.scanResultFooter(combinedReportResult));
    }

    private headingWithMessage = (message?: string): string => {
        if (message) {
            return heading(`${productTitle()}: ${message}`, 3);
        }
        return heading(`${productTitle()}`, 3);
    };

    private baselineDetails = (baselineInfo: BaselineInfo): string => {
        const baselineFileName = baselineInfo.baselineFileName;
        const baselineEvaluation = baselineInfo.baselineEvaluation;
        const baseliningDocsUrl = `https://github.com/microsoft/accessibility-insights-action/blob/main/docs/ado-extension-usage.md#using-a-baseline-file`;
        const baseliningDocsLink = link(baseliningDocsUrl, 'baselining docs');
        const scanArgumentsLink = link(this.artifactsInfoProvider.getArtifactsUrl(), 'scan arguments');
        const baselineNotConfiguredHelpText = `A baseline lets you mark known failures so it's easier to identify new failures as they're introduced. See ${baseliningDocsLink} for more.`;
        const baselineNotDetectedHelpText = `To update the baseline with these changes, copy the updated baseline file to ${scanArgumentsLink}. See ${baseliningDocsLink} for more.`;
        let lines = [''];

        if (baselineFileName === undefined) {
            lines = [bold('Baseline not configured'), sectionSeparator(), baselineNotConfiguredHelpText];
        } else if (baselineEvaluation === undefined) {
            lines = [bold('Baseline not detected'), sectionSeparator(), baselineNotDetectedHelpText];
        } else {
            const newFailures = baselineEvaluation.totalNewViolations;
            const baselineFailures = baselineEvaluation.totalBaselineViolations;
            if (baselineFailures === undefined || (baselineFailures === 0 && newFailures > 0)) {
                lines = [bold('Baseline not detected'), sectionSeparator(), baselineNotDetectedHelpText];
            } else if (baselineFailures > 0) {
                const headingWithBaselineFailures = `${baselineFailures} failure instances in baseline`;
                let baselineFailuresHelpText = `not shown; see ${baseliningDocsLink}`;
                if (newFailures > 0) {
                    baselineFailuresHelpText = baselineFailuresHelpText.concat(' for how to include new failures into the baseline');
                }
                lines = [bold(headingWithBaselineFailures), sectionSeparator(), `(${baselineFailuresHelpText})`];
            }
        }

        return lines.join('');
    };

    private urlsListItem = (passedUrls: number, unscannableUrls: number, failedUrls: number): string => {
        const failedUrlsSummary = `${failedUrls} URL(s) failed, `;
        const passedAndUnscannableUrlsSummary = `${passedUrls} URL(s) passed, and ${unscannableUrls} were not scannable`;
        const urlsSummary = failedUrls === 0 ? passedAndUnscannableUrlsSummary : failedUrlsSummary.concat(passedAndUnscannableUrlsSummary);
        return listItem(`${bold(`URLs`)}: ${urlsSummary}`);
    };

    private urlsListItemBaseline = (passedUrls: number, unscannableUrls: number, failedUrls: number): string => {
        const urlsSummary = `${failedUrls} with failures, ${passedUrls} passed, ${unscannableUrls} not scannable`;
        return `${bold(`URLs`)}: ${urlsSummary}`;
    };

    private rulesListItem = (passedChecks: number, inapplicableChecks: number, failedChecks: number) => {
        const failedRulesSummary = `${failedChecks} check(s) failed, `;
        const passedAndInapplicableRulesSummary = `${passedChecks} check(s) passed, and ${inapplicableChecks} were not applicable`;
        const rulesSummary =
            failedChecks === 0 ? passedAndInapplicableRulesSummary : failedRulesSummary.concat(passedAndInapplicableRulesSummary);
        return listItem(`${bold(`Rules`)}: ${rulesSummary}`);
    };

    private rulesListItemBaseline = (passedChecks: number, inapplicableChecks: number, failedChecks: number) => {
        const rulesSummary = `${failedChecks} with failures, ${passedChecks} passed, ${inapplicableChecks} not applicable`;
        return `${bold(`Rules`)}: ${rulesSummary}`;
    };

    private failureDetails = (combinedReportResult: CombinedReportParameters): string => {
        if (combinedReportResult.results.resultsByRule.failed.length === 0) {
            return '';
        }

        const failedRulesList = combinedReportResult.results.resultsByRule.failed.map((failuresGroup) => {
            const failureCount = failuresGroup.failed.length;
            const ruleId = failuresGroup.failed[0].rule.ruleId;
            const ruleDescription = failuresGroup.failed[0].rule.description;
            return [this.failedRuleListItem(failureCount, ruleId, ruleDescription), sectionSeparator()].join('');
        });
        const lines = [sectionSeparator(), `${heading('Failed instances', 4)}`, sectionSeparator(), ...failedRulesList];

        return lines.join('');
    };

    private failedRuleListItem = (failureCount: number, ruleId: string, description: string) => {
        return listItem(`${bold(`${failureCount} Ã— ${escaped(ruleId)}`)}:  ${escaped(description)}`);
    };

    private failureDetailsBaseline = (combinedReportResult: CombinedReportParameters, baselineInfo: BaselineInfo): string => {
        let lines = [];
        if (this.hasFailures(combinedReportResult, baselineInfo.baselineEvaluation)) {
            const failedRulesList = this.getFailedRulesList(combinedReportResult, baselineInfo.baselineEvaluation);
            const failureInstances = this.getFailureInstances(combinedReportResult, baselineInfo.baselineEvaluation);
            const failureInstancesHeading = this.getFailureInstancesHeading(failureInstances, baselineInfo.baselineEvaluation);
            lines = [sectionSeparator(), bold(failureInstancesHeading), sectionSeparator(), ...failedRulesList];
        } else {
            lines = this.getNoFailuresText(baselineInfo.baselineEvaluation);
        }

        return lines.join('');
    };

    private getNoFailuresText = (baselineEvaluation: BaselineEvaluation): string[] => {
        const checkMark = ':white_check_mark:';
        const pointRight = ':point_right:';
        let failureDetailsHeading = `${checkMark} No failures detected`;
        let failureDetailsDescription = `No failures were detected by automatic scanning.`;
        if (this.baselineHasFailures(baselineEvaluation)) {
            failureDetailsHeading = `${checkMark} No new failures`;
            failureDetailsDescription = 'No failures were detected by automatic scanning except those which exist in the baseline.';
        }
        const nextStepHeading = `${pointRight} Next step:`;
        const tabStopsUrl = `https://accessibilityinsights.io/docs/en/web/getstarted/fastpass/#complete-the-manual-test-for-tab-stops`;
        const tabStopsLink = link(tabStopsUrl, 'Accessibility Insights Tab Stops');
        const nextStepDescription = ` Manually assess keyboard accessibility with ${tabStopsLink}`;

        return [
            bold(failureDetailsHeading),
            sectionSeparator(),
            failureDetailsDescription,
            sectionSeparator(),
            sectionSeparator(),
            bold(nextStepHeading),
            nextStepDescription,
            sectionSeparator(),
        ];
    };

    private getTotalFailureInstancesFromResults = (combinedReportResult: CombinedReportParameters): number => {
        return combinedReportResult.results.resultsByRule.failed.reduce((a, b) => a + b.failed.reduce((c, d) => c + d.urls.length, 0), 0);
    };

    private getFailureInstances = (combinedReportResult: CombinedReportParameters, baselineEvaluation: BaselineEvaluation): number => {
        if (baselineEvaluation) {
            return baselineEvaluation.totalNewViolations;
        }

        return this.getTotalFailureInstancesFromResults(combinedReportResult);
    };

    private getFailureInstancesHeading = (failureInstances: number, baselineEvaluation: BaselineEvaluation): string => {
        let failureInstancesHeading = `${failureInstances} failure instances`;
        if (this.baselineHasFailures(baselineEvaluation)) {
            failureInstancesHeading = failureInstancesHeading.concat(' not in baseline');
        }

        return failureInstancesHeading;
    };

    private getFailedRulesList = (combinedReportResult: CombinedReportParameters, baselineEvaluation: BaselineEvaluation): string[] => {
        if (baselineEvaluation) {
            return this.getNewFailuresList(combinedReportResult, baselineEvaluation);
        }

        return this.getFailedRulesListWithNoBaseline(combinedReportResult);
    };

    private getNewFailuresList = (combinedReportResult: CombinedReportParameters, baselineEvaluation: BaselineEvaluation): string[] => {
        const newFailuresList = [];
        for (const ruleId in baselineEvaluation.newViolationsByRule) {
            const failureCount = baselineEvaluation.newViolationsByRule[ruleId];
            const ruleDescription = this.getRuleDescription(combinedReportResult, ruleId);
            newFailuresList.push([this.failedRuleListItemBaseline(failureCount, ruleId, ruleDescription), sectionSeparator()].join(''));
        }

        return newFailuresList;
    };

    private getFailedRulesListWithNoBaseline = (combinedReportResult: CombinedReportParameters): string[] => {
        const failedRulesList = combinedReportResult.results.resultsByRule.failed.map((failuresGroup) => {
            const failureCount = failuresGroup.failed.reduce((a, b) => a + b.urls.length, 0);
            const ruleId = failuresGroup.failed[0].rule.ruleId;
            const ruleDescription = failuresGroup.failed[0].rule.description;
            return [this.failedRuleListItemBaseline(failureCount, ruleId, ruleDescription), sectionSeparator()].join('');
        });

        return failedRulesList;
    };

    private getRuleDescription = (combinedReportResult: CombinedReportParameters, ruleId: string): string => {
        const matchingFailuresGroup = combinedReportResult.results.resultsByRule.failed.find((failuresGroup) => {
            return failuresGroup.failed[0].rule.ruleId === ruleId;
        });

        return matchingFailuresGroup.failed[0].rule.description;
    };

    private failedRuleListItemBaseline = (failureCount: number, ruleId: string, description: string) => {
        return listItem(`(${failureCount}) ${bold(escaped(ruleId))}:  ${escaped(description)}`);
    };

    private scanResultDetails(scanResult: string, footer?: string): string {
        const lines = [scanResult, sectionSeparator(), footerSeparator(), sectionSeparator(), footer];

        return lines.join('');
    }

    private scanResultFooter(combinedReportResult: CombinedReportParameters): string {
        const axeVersion = combinedReportResult.axeVersion;
        const axeCoreUrl = `https://github.com/dequelabs/axe-core/releases/tag/v${axeVersion}`;
        const axeLink = link(axeCoreUrl, `axe-core ${axeVersion}`);

        return `This scan used ${axeLink} with ${combinedReportResult.userAgent}.`;
    }

    private downloadArtifacts(): string {
        const artifactName = `${brand} artifact`;
        return listItem(`Download the ${bold(artifactName)} to view the detailed results of these checks`);
    }

    private downloadArtifactsWithLink(combinedReportResult: CombinedReportParameters, baselineEvaluation?: BaselineEvaluation): string {
        const artifactsLink = link(this.artifactsInfoProvider.getArtifactsUrl(), 'run artifacts');
        let details = 'all failures and scan details';
        if (!this.baselineHasFailures(baselineEvaluation) && !this.hasFailures(combinedReportResult, baselineEvaluation)) {
            details = 'scan details';
        }
        return `See ${details} by downloading the report from ${artifactsLink}`;
    }

    private baselineHasFailures = (baselineEvaluation: BaselineEvaluation): boolean => {
        return (
            baselineEvaluation !== undefined && baselineEvaluation.totalBaselineViolations && baselineEvaluation.totalBaselineViolations > 0
        );
    };

    private hasFailures = (combinedReportResult: CombinedReportParameters, baselineEvaluation: BaselineEvaluation): boolean => {
        if (baselineEvaluation !== undefined) {
            return baselineEvaluation.totalNewViolations > 0;
        }

        return this.getTotalFailureInstancesFromResults(combinedReportResult) > 0;
    };
}
