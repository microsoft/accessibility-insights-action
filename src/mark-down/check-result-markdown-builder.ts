// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AxeScanResults } from 'accessibility-insights-scan';
import * as axe from 'axe-core';
import { injectable } from 'inversify';

import { brand, brandLogoImg } from '../content/strings';
import { bold, footerSeparator, heading, link, listItem, productTitle, sectionSeparator } from '../utils/markdown-formatter';

@injectable()
export class CheckResultMarkdownBuilder {
    public failureDetails = (axeScanResults: AxeScanResults) => {
        const failedRulesList = axeScanResults.results.violations.map((rule: axe.Result) => {
            return [this.failedRuleListItem(rule.nodes.length, rule.id, rule.description), sectionSeparator()].join('');
        });
        const sections = [
            this.failureSummary(
                axeScanResults.results.violations.length,
                axeScanResults.results.passes.length,
                axeScanResults.results.inapplicable.length,
            ),
            sectionSeparator(),

            `${heading('Failed instances', 4)}`,
            sectionSeparator(),
        ];

        return this.scanResultDetails(sections.concat(failedRulesList).join(''), this.scanResultFooter(axeScanResults));
    };

    public errorContent = (): string => {
        const lines = [
            heading(`${productTitle()}: Something went wrong`, 3),
            sectionSeparator(),

            `You can review the log to troubleshoot the issue. Fix it and re-run the workflow to run the automated accessibility checks again.`,
            sectionSeparator(),
        ];

        return this.scanResultDetails(lines.join(''));
    };

    public congratsContent = (axeScanResults: AxeScanResults) => {
        const passed = axeScanResults.results.passes.length;
        const inapplicable = axeScanResults.results.inapplicable.length;
        const lines = [
            heading(`${productTitle()}: All applicable checks passed`, 3),
            sectionSeparator(),

            listItem(`${bold(`${passed} check(s) passed`)}, and ${inapplicable} were not applicable`),
            sectionSeparator(),

            this.downloadArtifacts(),
        ];

        return this.scanResultDetails(lines.join(''), this.scanResultFooter(axeScanResults));
    };

    private readonly scanResultDetails = (scanResult: string, footer?: string): string => {
        const lines = [scanResult, sectionSeparator(), footerSeparator(), sectionSeparator(), footer];

        return lines.join('');
    };

    private readonly scanResultFooter = (axeScanResults: AxeScanResults): string => {
        const axeVersion = axeScanResults.results.testEngine.version;
        const axeCoreUrl = `https://github.com/dequelabs/axe-core/releases/tag/v${axeVersion}`;
        const axeLink = link(axeCoreUrl, `axe-core ${axeVersion}`);

        return `This scan used ${axeLink} with ${axeScanResults.browserSpec}.`;
    };

    private readonly failureSummary = (failed: number, passed: number, inapplicable: number) => {
        const lines = [
            heading(`${productTitle()}`, 3),
            sectionSeparator(),

            listItem(`${bold(`${failed} check(s) failed`)}, ${passed} passed, and ${inapplicable} were not applicable`),
            sectionSeparator(),
            this.downloadArtifacts(),
        ];

        return lines.join('');
    };

    private readonly failedRuleListItem = (failureCount: number, ruleId: string, description: string) => {
        return listItem(`${bold(`${failureCount} Ã— ${ruleId}`)}:  ${description}`);
    };

    private readonly downloadArtifacts = () => {
        return listItem(`Download the ${bold(brand)} artifact to view the detailed results of these checks`);
    };
}
