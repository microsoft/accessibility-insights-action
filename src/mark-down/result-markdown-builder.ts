// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { injectable } from 'inversify';
import { brand } from '../content/strings';
import { bold, footerSeparator, heading, link, listItem, productTitle, sectionSeparator } from './markdown-formatter';
import { CombinedReportParameters } from 'accessibility-insights-report';

@injectable()
export class ResultMarkdownBuilder {
    public buildErrorContent(): string {
        const lines = [
            heading(`${productTitle()}: Something went wrong`, 3),
            sectionSeparator(),
            `You can review the log to troubleshoot the issue. Fix it and re-run the workflow to run the automated accessibility checks again.`,
            sectionSeparator(),
        ];

        return this.scanResultDetails(lines.join(''));
    }

    public buildContent(combinedReportResult: CombinedReportParameters): string {
        const passedChecks = combinedReportResult.results.resultsByRule.passed.length;
        const inapplicableChecks = combinedReportResult.results.resultsByRule.notApplicable.length;
        const failedChecks = combinedReportResult.results.resultsByRule.failed.reduce((a, b) => a + b.failed.length, 0);

        const lines = [
            heading(`${productTitle()}`, 3),
            sectionSeparator(),
            listItem(
                `${bold(
                    `Rules`,
                )}: ${failedChecks} check(s) failed, ${passedChecks} check(s) passed, and ${inapplicableChecks} were not applicable`,
            ),
            sectionSeparator(),
            listItem(
                `${bold(`URLs`)}: ${combinedReportResult.results.urlResults.failedUrls} URL(s) failed, ${
                    combinedReportResult.results.urlResults.passedUrls
                } URL(s) passed, and ${combinedReportResult.results.urlResults.unscannableUrls} were not scannable`,
            ),
            sectionSeparator(),

            this.downloadArtifacts(),
        ];

        return this.scanResultDetails(lines.join(''), this.scanResultFooter(combinedReportResult));
    }

    private scanResultDetails(scanResult: string, footer?: string): string {
        const lines = [scanResult, sectionSeparator(), footerSeparator(), sectionSeparator(), footer];

        return lines.join('');
    }

    private scanResultFooter(combinedReportResult: CombinedReportParameters): string {
        const axeVersion = combinedReportResult.axeVersion;
        const axeCoreUrl = `https://github.com/dequelabs/axe-core/releases/tag/v${axeVersion}`;
        const axeLink = link(axeCoreUrl, `axe-core ${axeVersion}`);

        return `This scan used ${axeLink} with ${combinedReportResult.userAgent}.`;
    }

    private downloadArtifacts(): string {
        return listItem(`Download the ${bold(brand)} artifact to view the detailed results of these checks`);
    }
}
