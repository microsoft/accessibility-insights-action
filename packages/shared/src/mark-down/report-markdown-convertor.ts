// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { inject, injectable } from 'inversify';
import { ResultMarkdownBuilder } from './result-markdown-builder';
import { CombinedReportParameters } from 'accessibility-insights-report';

@injectable()
export class ReportMarkdownConvertor {
    constructor(@inject(ResultMarkdownBuilder) private readonly checkResultMarkdownBuilder: ResultMarkdownBuilder) {}

    public convert(combinedReportResult: CombinedReportParameters): string {
        return this.checkResultMarkdownBuilder.buildContent(combinedReportResult);
    }

    public getErrorMarkdown(): string {
        return this.checkResultMarkdownBuilder.buildErrorContent();
    }
}
