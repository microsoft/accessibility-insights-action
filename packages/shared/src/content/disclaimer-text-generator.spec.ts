// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import 'reflect-metadata';
import { IMock, Mock } from 'typemoq';
import { MarkdownOutputFormatter } from '../mark-down/markdown-output-formatter';
import { DisclaimerTextGenerator } from './disclaimer-text-generator';

describe(DisclaimerTextGenerator, () => {
    let testSubject: DisclaimerTextGenerator;
    let markdownOutputFormatterMock: IMock<MarkdownOutputFormatter>;

    beforeEach(() => {
        markdownOutputFormatterMock = Mock.ofType<MarkdownOutputFormatter>();
        testSubject = new DisclaimerTextGenerator(markdownOutputFormatterMock.object);
    });

    it('generates disclaimer text', () => {
        const disclaimerText = testSubject.getDisclaimerText();
        expect(disclaimerText).toMatchSnapshot();
    });
});
