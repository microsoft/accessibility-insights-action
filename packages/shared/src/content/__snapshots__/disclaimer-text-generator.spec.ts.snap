// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisclaimerTextGenerator generates disclaimer text 1`] = `
"The Accessibility Insights Action runs a set of automated checks to help find some of the most common accessibility issues. The automated checks can detect accessibility problems such as missing or invalid properties, but most accessibility problems can only be discovered through manual testing.

We recommend automated testing, to continuously protect against some common issues, and regular undefined using undefined, a free and open source tool that helps you assess your website or web app for undefined coverage."
`;
